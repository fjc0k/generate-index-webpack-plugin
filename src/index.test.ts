import GenerateIndexPlugin from '.'
import tmp from 'tempy'
import webpack from 'webpack'
import { copySync, readFileSync, removeSync, writeFileSync } from 'fs-extra'
import { join } from 'path'

let cwd: string

beforeEach(() => {
  cwd = tmp.directory()
  copySync(join(__dirname, './__fixtures__'), join(cwd, 'src'))
})

afterEach(() => {
  removeSync(cwd)
})

test('run', done => {
  const compiler = webpack({
    entry: './src/index.js',
    output: {
      path: join(cwd, './dist'),
    },
    optimization: {
      minimize: false,
    },
    plugins: [
      new GenerateIndexPlugin({
        patterns: ['./src/index.js', './src/assets/index.js'],
      }),
    ],
    context: cwd,
  })
  compiler.run(err => {
    expect(err).toBeFalsy()
    expect(
      readFileSync(join(cwd, './src/index.js')).toString(),
    ).toMatchSnapshot('src/index.js')
    expect(
      readFileSync(join(cwd, './src/assets/index.js')).toString(),
    ).toMatchSnapshot('src/assets/index.js')
    expect(
      eval(readFileSync(join(cwd, './dist/main.js')).toString()),
    ).toMatchSnapshot('eval dist/main.js')
    done()
  })
})

test('watch-run', done => {
  const compiler = webpack({
    entry: './src/index.js',
    output: {
      path: join(cwd, './dist'),
    },
    optimization: {
      minimize: false,
    },
    plugins: [
      new GenerateIndexPlugin({
        patterns: ['./src/index.js', './src/assets/index.js'],
      }),
    ],
    context: cwd,
  })
  let lastHash: string | undefined
  let i = 0
  const watching = compiler.watch({}, (err, stats) => {
    if (stats.hash === lastHash) return
    lastHash = stats.hash
    i++
    expect(err).toBeFalsy()
    expect(
      readFileSync(join(cwd, './src/index.js')).toString(),
    ).toMatchSnapshot(`<${i}> src/index.js`)
    expect(
      readFileSync(join(cwd, './src/assets/index.js')).toString(),
    ).toMatchSnapshot(`<${i}> src/assets/index.js`)
    expect(
      eval(readFileSync(join(cwd, './dist/main.js')).toString()),
    ).toMatchSnapshot(`<${i}> eval dist/main.js`)
  })
  setTimeout(() => {
    writeFileSync(join(cwd, './src/hello.js'), 'export const hello = "ok"')
    setTimeout(() => {
      watching.close(() => {
        done()
      })
    }, 1000)
  }, 1000)
})
