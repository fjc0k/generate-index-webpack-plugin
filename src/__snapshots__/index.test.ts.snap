// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`run: eval dist/main.js 1`] = `
Object {
  "a": 100,
  "ddd": "dddd",
  "imgButtonRed": Object {},
  "imgHome": Object {},
  "imgLogo": Object {},
}
`;

exports[`run: src/assets/index.js 1`] = `
"// @index('./*', (f, _) => \`export { default as img\${_.pascalCase(f.name)} } from '\${f.path}\${f.ext}'\`)
export { default as imgButtonRed } from './button-red.jpg'
export { default as imgHome } from './home.png'
export { default as imgLogo } from './logo.svg'
// @endindex
"
`;

exports[`run: src/index.js 1`] = `
"// @index(['./*', '!./dist'], f => \`export * from '\${f.path}'\`)
export * from './_ddllds'
export * from './a'
export * from './assets'
// @endindex
"
`;

exports[`watch-run: <1> eval dist/main.js 1`] = `
Object {
  "a": 100,
  "ddd": "dddd",
  "imgButtonRed": Object {},
  "imgHome": Object {},
  "imgLogo": Object {},
}
`;

exports[`watch-run: <1> src/assets/index.js 1`] = `
"// @index('./*', (f, _) => \`export { default as img\${_.pascalCase(f.name)} } from '\${f.path}\${f.ext}'\`)
export { default as imgButtonRed } from './button-red.jpg'
export { default as imgHome } from './home.png'
export { default as imgLogo } from './logo.svg'
// @endindex
"
`;

exports[`watch-run: <1> src/index.js 1`] = `
"// @index(['./*', '!./dist'], f => \`export * from '\${f.path}'\`)
export * from './_ddllds'
export * from './a'
export * from './assets'
// @endindex
"
`;

exports[`watch-run: <2> eval dist/main.js 1`] = `
Object {
  "a": 100,
  "ddd": "dddd",
  "hello": "ok",
  "imgButtonRed": Object {},
  "imgHome": Object {},
  "imgLogo": Object {},
}
`;

exports[`watch-run: <2> src/assets/index.js 1`] = `
"// @index('./*', (f, _) => \`export { default as img\${_.pascalCase(f.name)} } from '\${f.path}\${f.ext}'\`)
export { default as imgButtonRed } from './button-red.jpg'
export { default as imgHome } from './home.png'
export { default as imgLogo } from './logo.svg'
// @endindex
"
`;

exports[`watch-run: <2> src/index.js 1`] = `
"// @index(['./*', '!./dist'], f => \`export * from '\${f.path}'\`)
export * from './_ddllds'
export * from './a'
export * from './assets'
export * from './hello'
// @endindex
"
`;
